lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -4089,8 +4089,13 @@ static void diagnoseConformanceFailure(Type T,
     }
   }
 
-  diags.diagnose(ComplainLoc, diag::type_does_not_conform,
-                 T, Proto->getDeclaredType());
+  diags.diagnose(ComplainLoc, diag::type_does_not_conform, T,
+                 Proto->getDeclaredType());
+  // Attempt to diagnose conformance failure because requirements for a
+  // conformance couldn't be automatically derived
+  if (auto *nominal = T->getNominalOrBoundGenericNominal()) {
```

Comment:
I might be totally wrong but I think this function is called when we check conformance of a type at the usage location.

I see Equatable/Hashable synthesis already has a restriction on being declared in the same file as the original type. This is checked by `checkAndDiagnoseDisallowedContext()` which is called from `DerivedConformance::deriveHashable()` and `DerivedConformance::deriveEquatable()`. Maybe we need to add the new checks there?

Reply:
I looked into this a little more, and ended up moving the checks to a `diagnoseOrDefer` block in `ConformanceChecker::resolveWitnessViaLookup` instead. This way, the new notes are emitted alongside the existing 'candidate would match' notes for each 'does not conform' error. This ended up being simpler then trying to integrate them into `checkAndDiagnoseDisallowedContext()`, which isn't called currently if it can be determined in advance conformance synthesis won't succeed. 

If this doesn't seem like a good idea I can definitely consider other possible strategies.